File: LinkedListOne.java
package LeetCode;
import java.util.*;
public class LinkedListOne {
	
	  public class ListNode {
	      int val;
	      ListNode next;
	     ListNode(int x) { val = x; }
	 }
	 
	
	//Reverse a linked List
	//TC : O(N)
	//SC : O(!)
	class Solution1 {
	    public ListNode reverseList(ListNode head) {
	        
	        if( head ==null || head.next == null)
	            return head;
	        
	         ListNode last = reverseList(head.next); 
	           head.next.next = head;
	           head.next = null;
	        return last;
	     }
	}
	
	
	//Remove Nth Node From End of List
	//TC : O(n)
	// SC : O(1)
	class Solution2 {
	    public ListNode removeNthFromEnd(ListNode head, int count) {
	        if(count ==1 && head == null || head.next == null)
	            return null;
	       
	        ListNode temp = head;
	        while(count > 0){
	            temp = temp.next;
	            count--;
	        }
	        ListNode slow = head;
	        while(temp!=null)
	        {
	            slow = slow.next;
	            temp=temp.next;
	        }
	        
	        temp = head;
	        
	        while(temp.next!=slow && temp!=slow)
	        temp = temp.next;
	        
	        if(temp == slow)
	         head = temp.next;
	        
	        temp.next = slow.next;
	        return head;
	    }
	}
	
	 //Linked List Cycle II
	//TC O(n)
	//SC O(n)
	public class Solution {
	    public ListNode detectCycle(ListNode head) {
	        HashSet<ListNode> set = new HashSet<>();
	        
	        while(head!=null){
	            if(set.contains(head))
	                return head;
	            set.add(head);
	            head=head.next;
	    
	        }
	        return head;
	    }
	}
	
	
}

File: LinkedListOne.py
# Linked-List-1

## Problem1 (https://leetcode.com/problems/reverse-linked-list/)
# Time Complexity : O(n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Iterative Solution
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head:
            return None
        prev = None
        curr = head
        fast = head.next
        while fast:
            curr.next = prev
            prev = curr
            curr = fast
            fast = fast.next
        curr.next = prev
        return curr

# Recursive Solution
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        newHead = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return newHead


## Problem2 (https://leetcode.com/problems/remove-nth-node-from-end-of-list/)
# Time Complexity : O(n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Two pass solution
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        if not head:
            return None
        dummy = ListNode(-1)
        dummy.next = head
        curr = head
        length = 0
        while curr:
            length += 1
            curr = curr.next
        length -= n
        curr = dummy
        while length > 0:
            length -= 1
            curr = curr.next
        curr.next = curr.next.next
        return dummy.next

## Problem3 (https://leetcode.com/problems/linked-list-cycle-ii/)
# Time Complexity : O(n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Floyd's Cycle Detection Algorithm
class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        if not head:
            return None
        slow = head
        fast = head
        flag = False
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                flag = True
                break
        if not flag:
            return None
        slow = head
        while slow != fast:
            slow = slow.next
            fast = fast.next
        return slow

